src/models/model.cpp

```diff
@@ -249,32 +249,149 @@ int32_t Tokenizer::TokenToTokenId(const char* token) const {
+      session_options.AppendExecutionProvider("WebGPU", opts);
+    } else
+      throw std::runtime_error("Unknown provider type: " + provider_options.name);
+  }
+  return p_device;
+}
+
+// Trivial ONNX model that just returns a single float constant. Used below to create an OrtSession that
+// lets us get a device Ort::Allocator for each device type. This is necessary because the Ort::Allocator
+// needs to persist and is valid for the lifetime of this OrtSession.
+static const uint8_t g_trivial_model[] = {
+    0x08, 0x0a, 0x12, 0x01, 0x61, 0x3a, 0x53, 0x0a, 0x38, 0x12, 0x06, 0x76, 0x61, 0x6c, 0x75, 0x65,
+    0x73, 0x22, 0x08, 0x43, 0x6f, 0x6e, 0x73, 0x74, 0x61, 0x6e, 0x74, 0x2a, 0x24, 0x0a, 0x05, 0x76,
+    0x61, 0x6c, 0x75, 0x65, 0x2a, 0x18, 0x08, 0x01, 0x10, 0x01, 0x42, 0x0c, 0x63, 0x6f, 0x6e, 0x73,
+    0x74, 0x5f, 0x74, 0x65, 0x6e, 0x73, 0x6f, 0x72, 0x4a, 0x04, 0x00, 0x00, 0x00, 0x00, 0xa0, 0x01,
+    0x04, 0x12, 0x01, 0x62, 0x62, 0x14, 0x0a, 0x06, 0x76, 0x61, 0x6c, 0x75, 0x65, 0x73, 0x12, 0x0a,
+    0x0a, 0x08, 0x08, 0x01, 0x12, 0x04, 0x0a, 0x02, 0x08, 0x01, 0x42, 0x04, 0x0a, 0x00, 0x10, 0x15};
+
 // Since Python/Others can and will hold onto a generator object past the model object's lifetime we need to ensure
 // the allocator used is not destroyed until last. This keeps the allocator around until exit, after all other memory
 // has been destroyed. Without this, we will crash in the Onnxruntime BFCArena code when deleting tensors due to the
 // arena already being destroyed.
-void EnsureDeviceOrtInit(OrtSession& session, DeviceType type) {
+void EnsureDeviceOrtInit(DeviceType type) {
   // CPU Allocator is a special case, it's not in the owned 'allocator_device_' table below so we handle it separately
   if (type == DeviceType::CPU)
     return;
 
-  auto& device = GetOrtGlobals()->allocator_device_[static_cast<int>(type)];
-  if (device)
+  auto& allocator = GetOrtGlobals()->device_allocators_[static_cast<int>(type)];
+  if (allocator.allocator_)
     return;
 
-  static const char* device_type_names[] = {"CPU (Not used, see above)", "Cuda", "DML", "WebGPU_Buffer", "QnnHtpShared"};
+  // Names for the device types used by 'SetProviderSessionOptions'
+  static const char* device_type_names[] = {"CPU (Not used, see above)", "cuda", "dml", "webgpu", "qnn"};
   static_assert(std::size(device_type_names) == static_cast<size_t>(DeviceType::MAX));
 
-  auto name = device_type_names[static_cast<int>(type)];
+  // Create an OrtSessionOptions and set the options to use the DeviceType we're using here
+  auto session_options = OrtSessionOptions::Create();
+  std::vector<Config::ProviderOptions> provider_options_list;
+  provider_options_list.emplace_back(Config::ProviderOptions{device_type_names[static_cast<int>(type)], {}});
+  SetProviderSessionOptions(*session_options, provider_options_list, true, false);
+  session_options->SetLogSeverityLevel(ORT_LOGGING_LEVEL_ERROR);  // Errors only here, as warnings are not useful to the user
```

Comment:
would be good if this was configurable. for FL we might need to get GenAI to provide a custom logger to ORT so all output goes to the same file, with a configurable severity level so we can debug issues as getting native stdout/stderr from C# when asp.net is involved is non-trivial.

completely separate change though so FYI and not expecting this PR to change.